core/directory.go

```diff
@@ -503,6 +503,32 @@ func (dir *Directory) WithFile(
 	return dir, nil
 }
 
+func (dir *Directory) WithFiles(
+	ctx context.Context,
+	destDir string,
+	src []*File,
+	permissions *int,
+	owner *Ownership,
+) (*Directory, error) {
+	dir = dir.Clone()
+
+	var err error
+	for _, file := range src {
+		dir, err = dir.WithFile(
```

Comment:
While this is fine today, I think we could potentially invert this - have `WithFile` call `WithFiles`.

We could have `mergeStates` take multiple source inputs, which would potentially be much more efficient (translating to one `MergeOp` instead of lots of individual ones). That said, this is a trickier optimization (and I think is also somewhat blocked on https://github.com/dagger/dagger/issues/6073).

Just a note, no action required here (but we could potentially leave a `TODO` comment for follow-up).